### PR TITLE
fix: check if response was successful in  IDTokenCredentials.refresh

### LINKED ISSUE #1249: KeyError when generating IDTokenCredentials with bad permissions
When trying to generate a IDTokenCredentials token, if for some reason there is some sort of permissions error, the library crashes with a KeyError instead of providing useful info to fix the error.

#### Environment details

  - OS: Ubuntu 22.04
  - Python version: 3.10.6
  - pip version: 23.0
  - `google-auth` version: 2.16.0

#### Steps to reproduce

```python
import google.auth
import google.auth.transport.requests
import google.oauth2.id_token
from google.auth import impersonated_credentials

target_principal = "some-bad-name@myprooject.iam.gserviceaccount.com"
target_audience = "https://some_url.appspot.com"

credentials, project = google.auth.default()

target_credentials = impersonated_credentials.Credentials(
    source_credentials=credentials,
    target_principal=target_principal,
    target_scopes=["https://www.googleapis.com/auth/cloud-platform"],
    lifetime=120,
)

credentials = impersonated_credentials.IDTokenCredentials(
    target_credentials, target_audience=target_audience
)
request = google.auth.transport.requests.Request()
credentials.refresh(request)
print(credentials.token)
```

And the traceback is
```
Traceback (most recent call last):
  File "/workspace/test.py", line 23, in <module>
    credentials.refresh(request)
  File "/usr/local/lib/python3.10/dist-packages/google/auth/impersonated_credentials.py", line 445, in refresh
    id_token = response.json()["token"]
KeyError: 'token'
```

### PR DESCRIPTION
**Pull request by charlielito**
Fixes #1249

### CODE REVIEW

**Review by firezdog — COMMENTED** [review]
I hope you can get this in soon!

**Comment by charlielito** [comment]
How can this get in?

File: google/auth/impersonated_credentials.py [hunk 1]
```diff
@@ -438,11 +438,19 @@ def refresh(self, request):
...
+            response = authed_session.post(
+                url=iam_sign_endpoint,
+                headers=headers,
+                data=json.dumps(body).encode("utf-8"),
+            )
+        finally:
+            authed_session.close()
+
+        if response.status_code != http_client.OK:
+            raise exceptions.TransportError(
```
Comment (arithmetic1728) [inline google/auth/impersonated_credentials.py:451]: Please change to `exceptions.RefreshError`

**Review by arithmetic1728 — COMMENTED** [review]
please add unit tests to cover the code change. This repo requires 100% code coverage for any PR. You can check it by running `python -m nox -s cover`. Thanks!

**Comment by arithmetic1728** [comment]
I will copy the fix from this PR and add unit tests, and create a PR, it seems I am not able to change this PR directly.

**Comment by arithmetic1728** [comment]
opened PR #1315

**Comment by arithmetic1728** [comment]
thank you @charlielito for your fix!

**Pull request closed without merging**